src/commands/dsyms/upload.ts

```diff
@@ -190,6 +201,17 @@ export class UploadCommand extends Command {
       .reduce((acc, nextSlice) => acc.concat(nextSlice), [])
   }
 
+  private resolveConfig = async () => {
+    try {
+      this.config = await parseConfigFile(this.config, this.configPath || DEFAULT_CONFIG_PATH)
+      setApiKeyAndSiteEnvVariablesFromConfig(this.config)
+    } catch (error) {
+      if (this.configPath) {
+        throw error
+      }
+    }
+  }
```

Comment:
This control flow looks quite complex: no matter of command parameters, we always try to parse either user config (`this.configPath`) or a default one (`DEFAULT_CONFIG_PATH`) and then only capture errors when user config is configured (otherwise, ignore errors).

Took me a while, but I understand what's happening here and the final behaviour seems correct. However, given that we do similar logic in callee `parseConfigFile`, it doesn't seem very straightforward:
```ts
export const parseConfigFile = async <T>(baseConfig: T, configPath?: string): Promise<T> => {
  try {
    const resolvedConfigPath = configPath ?? 'datadog-ci.json'
    const parsedConfig = await getConfig(resolvedConfigPath)

    return deepExtend(baseConfig, parsedConfig)
  } catch (e) {
    if (e.code === 'ENOENT' && configPath) {
      throw new Error('Config file not found')
    }

    if (e instanceof SyntaxError) {
      throw new Error('Config file is not correct JSON')
    }
  }

  return baseConfig
}
```

Is it JS/TS convention to use Errors to control the program flow and ignore them conditionally? Coming from static languages background, it looks very fragile to me.  I'd find it much more readable (from top to bottom) if we check in the upstream logic if config resolution is needed:
```ts
if (this.configPath) {
   // EVERY error thrown from here would indicate user error, unconditionally:
   resolveConfig(this.configPath)

} else if (pathExists(DEFAULT_CONFIG_PATH)) {
   // EVERY error thrown from here would indicate user error, unconditionally:
   resolveConfig(DEFAULT_CONFIG_PATH)

} // OTHERWISE, user clearly doesn't want to use config file
```

Reply:
It might also be nice if more of this logic was pulled into `utils.ts`. The default config location won't change from command to command and the result should always be to call `setApiKeyAndSiteEnvVariablesFromConfig` anyway yes?

Reply:
It totally makes sense to move this logic into `utils.ts`. I think we can improve the error flow indeed, but it is common to use errors to control the flow in functional programming, it is sometimes referred as [railway programming](https://fsharpforfunandprofit.com/rop/)

Reply:
Good idea to put extract most of it into utils (that would solve the problem of duplicating constant for default config path) 👍.

While I agree that conditioning Errors makes total sense in FP, here we use it in imperative flow with side effects, which doesn't make the concept easier.